### ISSUE TITLE
Possible bug in nimbleRcall

### ISSUE THREAD

**Issue by perrydv** [nimble-dev/nimble#873]
A nimble-users post on April 12 suggests a possible bug in nimbleRcall.  Further investigation is needed.

**Comment by perrydv** [comment]
I can provide an update but not a fix to this bug.  This was raised in the nimble-users post using a `nimbleRcall` to `dcomp` in package `CompGLM`.  The problem does not seem to be about `PROTECT` as initially thought.  Instead it seems to be about the state of the RNG.  The same random number is used multiple times.

I think the problem is triggered because `dcomp` calls code compiled as an `Rcpp::export`, which automatically does a `GetRNGstate`-`PutRNGstate` pair.  One can see that `Rcpp` in the past year has updated how this is managed due to previous problems (see [this blog post](https://developer.r-project.org/Blog/public/2019/03/28/use-of-c---in-packages/)), which presumably means it is now done well.  There are descriptions of a similar kind of bug as we have occurring previously, but I have yet to find a clear explanation, although it must be out there.

We use `GetRNGstate` and `PutRNGstate`  in a simple way (in every generated `SEXP` function), which has not caused problems in the past.  I see that we should use `PutRNGstate` before going back to R in `nimbleRcall` and then `GetRNGstate` upon entry back into our generated C++.  However, when I tried inserting those, the problem persisted.  I also tried updating everything: R, Rcpp, and building dcomp from source.  Those steps didn't work.

We can at least tell users that the problem seems to be limited (I hope) to the case when nimbleRcall calls an R function that calls C code that uses the RNG states.  The problem might be further limited to C functions exposed via `Rcpp::export`, but I'm not sure.  This will take more investigation.

**Comment by pistacliffcho** [comment]
Hi Perry,

Do you think you can create a minimal example? Since I use Rcpp quite a bit, I thought I’d take a look.

So far, it looks like both nimble *and* Rcpp handle the rng issue in the exact manner I’m surprised this is coming up.

Best,
Cliff

**Comment by perrydv** [comment]
Hi @pistacliffcho .  Thanks.  It would be great if you can take a look.  Here is a minimal example.

Issue873_minimal_example.cpp:
```
#include <Rcpp.h>
using namespace Rcpp;

// [[Rcpp::export]]
double plus_one(double x) {
  return(x + 1);
}
```

In R:
```
library(Rcpp)

## Create an R function "plus_one" that calls C via Rcpp::export
sourceCpp("Issue873_minimal_example.cpp")

library(nimble)

## Create a nimbleFunction that wraps plus_one
nrc_plus_one <- nimbleRcall(
  function(x = double()) {},
  Rfun =  'plus_one',
  returnType = double()
)

## Create a nimbleFunction to be compiled that calls nrc_plus_one
use_plus_one <- nimbleFunction(
  run = function(x = double()) {
    z <- rnorm(1)  ## Draw one random number
    cat("first draw: ", z, "\n")
    ans <- nrc_plus_one(x) ## Do the plus_one
    z <- rnorm(1)  ## Draw another random number
    cat("second draw: ", z, "\n")
    return(ans)
    returnType(double())
  }
)

C_use_plus_one <- compileNimble(use_plus_one)
C_use_plus_one(1) ## This show two identical draws from rnorm. The RNG was incorrectly reset to its previous value.

## Next we see that if the nimbleRcall does not call to 
## Rcpp-compiled code, the RNG is (correctly) not reset to a previous value.

## Use a pure R version of plus_one
R_plus_one <- function(x) x + 1

## Wrap R_plus_one in a nimbleFunction
nrc_R_plus_one <- nimbleRcall(
  function(x = double()) {},
  Rfun =  'R_plus_one',
  returnType = double()
)

## Use nrc_R_plus_one in compiled nimbleFunction
use_R_plus_one <- nimbleFunction(
  run = function(x = double()) {
    z <- rnorm(1)
    cat("first draw: ", z, "\n")
    ans <- nrc_R_plus_one(x)
    z <- rnorm(1)
    cat("second draw: ", z, "\n")
    return(ans)
    returnType(double())
  }
)

C_use_R_plus_one <- compileNimble(use_R_plus_one)
C_use_R_plus_one(1) ## This show the two draws are not identical
```

If you use `nimbleOptions(pauseAfterWritingFiles)`, you can modify the generated C++ before using 'c' in the browser to continue with compilation.  While paused, you can find the two functions that construct a `LANGSXP` and use `Rf_eval` to call the R evaluator.  In general, I think we should use `PutRNGstate()` before calling `Rf_eval` and `GetRNGstate()` after, but doing so doesn't seem to solve the problem.  Note that in this example, the C++ called via Rcpp does not actually make any random draws, but I believe `Rcpp::export` encases any function with `GetRNGstate()` and `PutRNGstate()`.  nimble does the same, but I think `Rcpp::export` does it by an object that calls `GetRNGstate()` on construction and `PutRNGstate()` on destruction.

**Comment by pistacliffcho** [comment]
Well, at the very least, I can show this can be shown to be a pure Rcpp issue, although nimble may have a pure nimble version of it too. In short, if you do the exact same code, but make “use_plus_one” an Rcpp function, you get the same issue. See below. 

I’m going to look into whether there is a pure nimble version of the issue. I suspect there is, because as far as I can tell, nimble and Rcpp wrap functions with identical get/setRNG code. After digging around a little more, I'll post the pure Rcpp issue to Dirk et al. 


Rcpp code:

// [[Rcpp::export]]
void callMe(){
  // dumb joke
  Rcout << "Anytime";
}

// [[Rcpp::export]]
void simAndCall(Function fun, bool callRFun){
  // fun is any R function, even one compiled by Rcpp
  double val = R::rnorm(0.0, 1.0);
  Rcout << "Val = " << val << "\n";
  if(callRFun){
    SEXP jnk = fun();
  }
  val = R::rnorm(0.0, 1.0);
  Rcout << "\nVal = " << val << "\n";
}

After compiling, call from R: 

> simAndCall(callMe, TRUE)
Val = 0.130605
Anytime
Val = 0.130605
> simAndCall(callMe, FALSE)
Val = -0.00385473

Val = -1.64165

**Comment by pistacliffcho** [comment]
Okay, there is a pure nimble version of the same issue. In Perry’s example, if we sub out the Rcpp function with a Nimble function, we get the same issue. See code below. 

I’m going to post the issue to Rcpp and see if they have a quick answer. 


nimPlusOne = nimbleFunction(
  run = function(x = double()){
    ans = x + 1
    return(ans)
    returnType(double())
  }
)

npo = compileNimble(nimPlusOne)

nrc_plus_one <- nimbleRcall(
  function(x = double()) {},
  Rfun =  'npo',
  returnType = double()
)

## Create a nimbleFunction to be compiled that calls nrc_plus_one
use_plus_one <- nimbleFunction(
  run = function(x = double()) {
    z <- rnorm(1)  ## Draw one random number
    cat("first draw: ", z, "\n")
    ans <- nrc_plus_one(x) ## Do the plus_one
    z <- rnorm(1)  ## Draw another random number
    cat("second draw: ", z, "\n")
    return(ans)
    returnType(double())
  }
)

C_use_plus_one <- compileNimble(use_plus_one)
C_use_plus_one(1) ## This show two identical draws from rnorm. The RNG was incorrectly reset to its previous value.

**Comment by pistacliffcho** [comment]
Final email on the subject. 

I did not post the issue to Rcpp, as it seems to be a known issue that Dirk strongly objects to fixing <https://github.com/RcppCore/Rcpp/issues/823>. Basically, Dirk’s point is that if you are calling Rcpp code from Rcpp, you should call the pure C++ function, not a call to an R-wrapper around the C++ function. I would say that in those contexts, he has a good point, but that nimble is the counter example, unless we can get nimble to talk nicely to Rcpp C++ code. 

Anyways, here’s the best solution that I’ve come up with. Basically, if you want to be extra safe when calling an R function (i.e., protect against it being either a nimble or Rcpp function), you could use the following format:

run = function(){
GetRNGstate()
…
PutRNGstate()
R_fxn()
GetRNGsate()
…
PutRNGsate()
}

As a more explicit example, here’s Perry’s earlier example (with the Rcpp function replaced with a nimble function), but you have the option to surround the R function with Get/PutRNGstate()’s. 


> library(nimble)
> 
> nimPlusOne = nimbleFunction(
+   run = function(x = double()){
+     ans = x + 1
+     return(ans)
+     returnType(double())
+   }
+ )
> 
> npo = compileNimble(nimPlusOne)
compiling... this may take a minute. Use 'showCompilerOutput = TRUE' to see C++ compiler details.
compilation finished.
> 
> nrc_plus_one <- nimbleRcall(
+   function(x = double()) {},
+   Rfun =  'npo',
+   returnType = double()
+ )
> 
> ## Best solution I can think of: surround R function with 
> ## *pairs* of Get/PutRNGState() on both sides R function
> use_plus_one <- nimbleFunction(
+   run = function(x = double(), 
+                  surroundRcode = integer()) {
+     if(surroundRcode){ GetRNGstate(); }
+     z <- rnorm(1)  ## Draw one random number
+     cat("first draw: ", z, "\n")
+     
+     if(surroundRcode){ PutRNGstate();}
+     ans <- nrc_plus_one(x) ## Do the plus_one
+     if(surroundRcode){ GetRNGstate(); }
+     z <- rnorm(1)  ## Draw another random number
+     cat("second draw: ", z, "\n")
+     if(surroundRcode){ PutRNGstate();}
+     return(ans)
+     returnType(double())
+   }
+ )
Warning message:
For this nimbleFunction to compile, these objects must be defined as nimbleFunctions, nimbleFunctionLists, or nimbleLists before compilation: GetRNGstate, PutRNGstate. 
> 
> C_use_plus_one <- compileNimble(use_plus_one)
compiling... this may take a minute. Use 'showCompilerOutput = TRUE' to see C++ compiler details.
compilation finished.
> # Calling function in which we **don't** surround R function 
> # with Get/PutRNG
> C_use_plus_one(1, 0)
first draw: -0.529596
second draw: -0.529596
[1] 2
> # Calling function in which we do surround R function 
> # with Get/PutRNG
> C_use_plus_one(1, 1)
first draw: -1.653
second draw: 1.80382
[1] 2

**Comment by perrydv** [comment]
Thanks @pistacliffcho .

I agree it looks like Rcpp has set up a system that makes sense and that they are happy with.  Given the amount of traffic they handle, let's hold off on pinging them unless we see a deeper need to.

I think I had, in effect, tried your solution (which I tried to describe), and it appeared *not* to work.  Where the `GetRNGstate()` and `PutRNGstate()` appeared was different than what you showed, I thought the net effect should be the same.  I will put together that case again.

**Comment by perrydv** [comment]
I've set up a modified version of `nimbleRcall()` to include `GetRNGstate()` and `PutRNGstate()`.  It fixes the problem, both in the minimal example shown above and in the original user's case.  When I tried it Friday by directly modifying generated C++ to check the idea, it didn't seem to work.  There must have been something wrong with my workflow.  In any case, it works now. 

For the record, in nimble, the generated C functions called via `.Call` from R, which have `CALL` in the C function name, always use `GetRNGstate()` before calling the "real" C function and then `PutRNGstate()` after. This is done regardless of whether any random numbers will in fact be generated.  This is similar (but simpler) to Rcpp's `//[[Rcpp::export]]` system, which always accomplishes the same (by different implementation), and also does so regardless of whether any random numbers will be generated. 

The problem was that a `nimbleRcall` calls the R evaluator from C++, and it neglected to do a `PutRNGstate()` before passing control to the R evaluator and then `GetRNGstate()` after processing returns from the R evaluator.  Inserting those calls has fixed the problem.  We can't assume that the `nimbleRcall` will or will not use the RNG, so we must defensively include these calls.  In fact, the Rcpp code being called in these examples does not generate random numbers, but simply getting and then putting the RNG state was sufficient to reveal the problem.

I will file a PR next.
